ClassCastException when receiving SAML LogoutResponse

Hi,

I'm using Pac4j with Play Framework 2.8 using the latest 4.0.0 release (play-pac4j and pac4j-saml-opensamlv3)

I'm seeing a ClassCastException when receiving a LogoutResponse from my IDP:

Caused by: java.lang.ClassCastException: org.opensaml.saml.saml2.core.impl.LogoutResponseImpl cannot be cast to org.opensaml.saml.saml2.core.Response
	at org.pac4j.saml.profile.impl.AbstractSAML2MessageReceiver.receiveMessage(AbstractSAML2MessageReceiver.java:61)
	at org.pac4j.saml.logout.impl.SAML2LogoutProfileHandler.receive(SAML2LogoutProfileHandler.java:36)
	at org.pac4j.saml.credentials.extractor.SAML2CredentialsExtractor.extract(SAML2CredentialsExtractor.java:59)
	at org.pac4j.core.client.BaseClient.retrieveCredentials(BaseClient.java:61)
	at org.pac4j.core.client.IndirectClient.getCredentials(IndirectClient.java:140)
	at org.pac4j.core.engine.DefaultCallbackLogic.perform(DefaultCallbackLogic.java:81)
	at org.pac4j.play.CallbackController.lambda$callback$0(CallbackController.java:54)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
	... 7 common frames omitted


It looks like this regression was caused by this merge:

https://github.com/pac4j/pac4j/commit/d1c10240eabd8a3343bfd05ab79a435e08cee141#diff-3276716761b33bc5ca73d61859d97119

I'm fairly new to this project, but looks like all received messages go though AbstractSML2MessageReceiver and the assumption from the merge is that all objects passing through here are Reponse objects, which in this case they are not.

--- a/pac4j-saml/src/main/java/org/pac4j/saml/context/SAML2MessageContext.java
+++ b/pac4j-saml/src/main/java/org/pac4j/saml/context/SAML2MessageContext.java
@@ -13,6 +13,7 @@ import org.opensaml.saml.common.messaging.context.SAMLSubjectNameIdentifierConte
 import org.opensaml.saml.saml2.core.Assertion;
 import org.opensaml.saml.saml2.core.BaseID;
 import org.opensaml.saml.saml2.core.Response;
+import org.opensaml.saml.saml2.core.StatusResponseType;
 import org.opensaml.saml.saml2.core.SubjectConfirmation;
 import org.opensaml.saml.saml2.metadata.AssertionConsumerService;
 import org.opensaml.saml.saml2.metadata.IDPSSODescriptor;
@@ -123,7 +124,7 @@ public class SAML2MessageContext {
         return getSPAssertionConsumerService(spssoDescriptor, spssoDescriptor.getAssertionConsumerServices());
     }
 
-    public AssertionConsumerService getSPAssertionConsumerService(final Response response) {
+    public AssertionConsumerService getSPAssertionConsumerService(final StatusResponseType response) {
         final SPSSODescriptor spssoDescriptor = getSPSSODescriptor();
         final List<AssertionConsumerService> services = spssoDescriptor.getAssertionConsumerServices();
 
@@ -134,7 +135,7 @@ public class SAML2MessageContext {
                     return service;
                 }
             }
-            throw new SAMLException("Assertion consumer service with sdestination " + response.getDestination()
+            throw new SAMLException("Assertion consumer service with destination " + response.getDestination()
                 + " could not be found for spDescriptor " + spssoDescriptor);
         }
 

--- a/pac4j-saml/src/main/java/org/pac4j/saml/context/SAML2MessageContext.java
+++ b/pac4j-saml/src/main/java/org/pac4j/saml/context/SAML2MessageContext.java
@@ -12,7 +12,6 @@ import org.opensaml.saml.common.messaging.context.SAMLSelfEntityContext;
 import org.opensaml.saml.common.messaging.context.SAMLSubjectNameIdentifierContext;
 import org.opensaml.saml.saml2.core.Assertion;
 import org.opensaml.saml.saml2.core.BaseID;
-import org.opensaml.saml.saml2.core.Response;
 import org.opensaml.saml.saml2.core.StatusResponseType;
 import org.opensaml.saml.saml2.core.SubjectConfirmation;
 import org.opensaml.saml.saml2.metadata.AssertionConsumerService;

--- a/pac4j-saml/src/main/java/org/pac4j/saml/profile/impl/AbstractSAML2MessageReceiver.java
+++ b/pac4j-saml/src/main/java/org/pac4j/saml/profile/impl/AbstractSAML2MessageReceiver.java
@@ -4,7 +4,6 @@ import org.opensaml.saml.common.SAMLObject;
 import org.opensaml.saml.common.messaging.context.SAMLBindingContext;
 import org.opensaml.saml.common.messaging.context.SAMLPeerEntityContext;
 import org.opensaml.saml.common.xml.SAMLConstants;
-import org.opensaml.saml.saml2.core.Response;
 import org.opensaml.saml.saml2.core.StatusResponseType;
 import org.opensaml.saml.saml2.metadata.AssertionConsumerService;
 import org.opensaml.saml.saml2.metadata.EntityDescriptor;

--- a/pac4j-saml/src/main/java/org/pac4j/saml/profile/impl/AbstractSAML2MessageReceiver.java
+++ b/pac4j-saml/src/main/java/org/pac4j/saml/profile/impl/AbstractSAML2MessageReceiver.java
@@ -5,6 +5,7 @@ import org.opensaml.saml.common.messaging.context.SAMLBindingContext;
 import org.opensaml.saml.common.messaging.context.SAMLPeerEntityContext;
 import org.opensaml.saml.common.xml.SAMLConstants;
 import org.opensaml.saml.saml2.core.Response;
+import org.opensaml.saml.saml2.core.StatusResponseType;
 import org.opensaml.saml.saml2.metadata.AssertionConsumerService;
 import org.opensaml.saml.saml2.metadata.EntityDescriptor;
 import org.opensaml.saml.saml2.metadata.IDPSSODescriptor;
@@ -59,8 +60,11 @@ public abstract class AbstractSAML2MessageReceiver implements SAML2MessageReceiv
         decodedCtx.getSAMLBindingContext().setRelayState(relayState);
         context.getSAMLBindingContext().setRelayState(relayState);
 
-        final AssertionConsumerService acsService
-            = context.getSPAssertionConsumerService((Response) decodedCtx.getMessageContext().getMessage());
+        final StatusResponseType response = (StatusResponseType) decodedCtx.getMessageContext().getMessage();
+        if (response == null) {
+            throw new SAMLException("Response from the context cannot be null");
+        }
+        final AssertionConsumerService acsService = context.getSPAssertionConsumerService(response);
         decodedCtx.getSAMLEndpointContext().setEndpoint(acsService);
 
         final EntityDescriptor metadata = context.getSAMLPeerMetadataContext().getEntityDescriptor();

